tests/stdlib/tlists.nim

```diff
@@ -81,3 +81,88 @@ block tlistsToString:
     l.append('2')
     l.append('3')
     doAssert $l == """['1', '2', '3']"""
+
+block toSinglyLinkedList:
+  let l = seq[int].default
+  doAssert l.toSinglyLinkedList.toSeq == []
+  doAssert [1].toSinglyLinkedList.toSeq == [1]
+  doAssert [1, 2, 3].toSinglyLinkedList.toSeq == [1, 2, 3]
+
+block toDoublyLinkedList:
+  let l = seq[int].default
+  doAssert l.toDoublyLinkedList.toSeq == []
+  doAssert [1].toDoublyLinkedList.toSeq == [1]
+  doAssert [1, 2, 3].toDoublyLinkedList.toSeq == [1, 2, 3]
+
+block copy:
+  doAssert array[0, int].default.toSinglyLinkedList.copy.toSeq == []
+  doAssert [1].toSinglyLinkedList.copy.toSeq == [1]
+  doAssert [1, 2].toSinglyLinkedList.copy.toSeq == [1, 2]
+  doAssert [1, 2, 3].toSinglyLinkedList.copy.toSeq == [1, 2, 3]
+  type Foo = ref object
+    x: int
+  var f = Foo(x: 1)
+  let a = [f, f].toSinglyLinkedList
+  f.x = 42
+  assert a.head.value == f
+  assert a.head.next.value == f
+
+block add:
+  block: # SinglyLinkedList
+    block:
+      var
+        l0 = initSinglyLinkedList[int]()
+        l1 = [1].toSinglyLinkedList
+        l2 = [2, 3].toSinglyLinkedList
+        l3 = [4, 5, 6].toSinglyLinkedList
+        l4 = l3.copy
+        l5 = l3.copy
+      l0.add(l3)
+      l1.add(l4)
+      l2.add(l5)
+      doAssert l0.toSeq == [4, 5, 6]
+      doAssert l1.toSeq == [1, 4, 5, 6]
+      doAssert l2.toSeq == [2, 3, 4, 5, 6]
+      doAssert l3.toSeq == []
+      l2.add(l3) # re-adding l3 that was destroyed is now a no-op
+      doAssert l2.toSeq == [2, 3, 4, 5, 6]
+      doAssert l3.toSeq == []
+    block:
+      var
+        l0 = initSinglyLinkedList[int]()
+        l1 = [1].toSinglyLinkedList
+        l2 = [2, 3].toSinglyLinkedList
+        l3 = [4, 5, 6].toSinglyLinkedList
+      l3.add(l0)
+      l2.add(l1)
+      doAssert l3.toSeq == [4, 5, 6]
+      doAssert l2.toSeq == [2, 3, 1]
+  block: # DoublyLinkedList
```

Comment:
there's a ton of duplication here. instead use this:
```nim
template testCommon(initList, toList) =
  var
    l0 = initDoublyLinkedList[int]()
    l1 = [1].toList
    l2 = [2, 3].toList
    l3 = [4, 5, 6].toList
    l4 = l3.copy
    l5 = l3.copy
  l0.add(l3)
  l1.add(l4)
  l2.add(l5)

testCommon(initDoublyLinkedList, toDoublyLinkedList)
testCommon(initSinglyLinkedList, toSinglyLinkedList)
```

it guarantees coverage.

tests specific to doubly or singly can be in a different block

Reply:
Thanks, that will improve the code a lot, I was also bothered by the code duplication, but templates didn't occur to me.